FAQ Tip: ZFS to compress workload outputs (#880)
firesim/firesim: docs/Advanced-Usage/FAQs.rst

likely a bug in one of your custom bridge implementations or in FireSim. In
fact, for a given target-design, enabling printf synthesis, assertion synthesis,
autocounter, or Auto ILA, should not change the simulated behavior of the machine.

Is there a way to compress workload results when copying back to the manager instance?
--------------------------------------------------------------------------------------

FireSim doesn't support compressing workload results before copying them back to the manager instance.
Instead we recommend that you use a modern filesystem (like ZFS) to provide compression for you.
For example, if you want to use ZFS to transparently compress data:

#. Attach a new volume to your EC2 instance (either at runtime or during launch).
   This is where data will be stored in a compressed format.
#. Make sure that the volume is attached (using something like ``lsblk -f``).
   This new volume should not have a filesystem type and should be unmounted (named something like ``nvme1n1``).

> #. Install ZFS according to https://zfsonlinux.org/

Reviewer: This should link directly to the CentOS installation procedures.
```suggestion
#. Install ZFS according to <https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html>.
   Check `/etc/redhat-release` to verify the CentOS version of the manager instance.
```

The current AMI is based on CentOS 7.9.  The instructions I used were:
```sh
yum install https://zfsonlinux.org/epel/zfs-release.el7_9.noarch.rpm
rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
yum install epel-release kernel-devel zfs
```

#. Mount the volume and setup the ZFS filesystem with compression. 

> 

Reviewer: I would add a prominent warning against blindly copy-pasting `/dev/nvme1n1` without verification.

```suggestion
.. warning::
    Creating the zpool will destroy all pre-existing data on that partition.  Double-check that the device node is correct before running any commands.
```

::

>     zpool create <MOUNTPOINT> /dev/nvme1n1
>     zpool list
>     zfs set compression=gzip <MOUNTPOINT>

Reviewer: The dataset properties can be specified in one `zpool` command:
```suggestion
    zpool create -o ashift=12 -O compression=on <POOL_NAME> /dev/nvme1n1
    zpool list
    zfs list
```

Also, `compression=on` selects a reasonable default algorithm, usually `lz4`.  `gzip` is usually not recommended as it tends to have a worse trade-off between decompression speed and compression ratio.

> #. At this point, you can use the ``<MOUNTPOINT>`` area as a normal folder to store data into where it will

Reviewer: Conceptually, the pool and dataset names determine the default mount point, not the other way around.  It's possible to specify an alternative mount point for individual datasets with the `mountpoint` property.

```suggestion
#. At this point, you can use the ``/<POOL_NAME>`` a normal directory to store data into where it will
```

FireSim doesn't support compressing workload results before copying them back to the manager instance.
Instead we recommend that you use a modern filesystem (like ZFS) to provide compression for you.
For example, if you want to use ZFS to transparently compress data:

#. Attach a new volume to your EC2 instance (either at runtime or during launch).
   This is where data will be stored in a compressed format.
#. Make sure that the volume is attached (using something like ``lsblk -f``).
   This new volume should not have a filesystem type and should be unmounted (named something like ``nvme1n1``).
#. Install ZFS according to <https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html>.
   Check ``/etc/redhat-release`` to verify the CentOS version of the manager instance.
#. Mount the volume and setup the ZFS filesystem with compression.

.. warning::
    Creating the zpool will destroy all pre-existing data on that partition.  Double-check that the device node is correct before running any commands.
::

    zpool create -o ashift=12 -O compression=on <POOL_NAME> /dev/nvme1n1
    zpool list
    zfs list

> #. At this point, you can use ``/<POOL_NAME>`` as a normal directory to store data into where it will

Reviewer: Does the list numbering continue properly for the last item?  Otherwise this can be formatted as  a paragraph.

Author: Ok fixed. Verified locally.

![image](https://user-images.githubusercontent.com/8823803/142508125-f78542fe-d8a1-483e-905d-9387ad55e08e.png)
